src/core/thread/link_metrics_tlvs.hpp

```diff
@@ -0,0 +1,332 @@
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void)
+    {
+        SetType(kLinkMetricsReportSub);
+        SetLength(sizeof(*this) - sizeof(Tlv));
+    }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval true   The TLV appears to be well-formed.
+     * @retval false  The TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(Tlv); }
+
+    /**
+     * This method returns the Link Metrics Type ID.
+     *
+     * @returns The Link Metrics Type ID.
+     *
+     */
+    LinkMetricsTypeIdFlags GetMetricsTypeId(void) const { return mMetricsTypeId; }
+
+    /**
+     * This method sets the Link Metrics Type ID.
+     *
+     * @param[in]  aMetricsTypeID  The Link Metrics Type ID to set.
+     *
+     */
+    void SetMetricsTypeId(LinkMetricsTypeIdFlags aMetricsTypeId)
+    {
+        mMetricsTypeId = aMetricsTypeId;
+        if (!aMetricsTypeId.IsLengthFlagSet())
+        {
+            SetLength(sizeof(*this) - sizeof(Tlv) - sizeof(uint32_t) + sizeof(uint8_t)); // The value is 1 byte long
+        }
+    }
+
+    /**
+     * This method returns the metric value in 8 bits.
+     *
+     * @returns The metric value.
+     *
+     */
+    uint8_t GetMetricsValue8(void) const { return mMetricsValue.m8; }
+
+    /**
+     * This method returns the metric value in 32 bits.
+     *
+     * @returns The metric value.
+     *
+     */
+    uint32_t GetMetricsValue32(void) const { return mMetricsValue.m32; }
+
+    /**
+     * This method sets the metric value (8 bits).
+     *
+     * @param[in]  aMetricsValue  Metrics value.
+     *
+     */
+    void SetMetricsValue8(uint8_t aMetricsValue) { mMetricsValue.m8 = aMetricsValue; }
+
+    /**
+     * This method sets the metric value (32 bits).
+     *
+     * @param[in]  aMetricsValue  Metrics value.
+     *
+     */
+    void SetMetricsValue32(uint32_t aMetricsValue) { mMetricsValue.m32 = aMetricsValue; }
+
+private:
+    LinkMetricsTypeIdFlags mMetricsTypeId;
+    union
+    {
+        uint8_t  m8;
+        uint32_t m32;
+    } mMetricsValue;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Link Metrics Query Options Sub-TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class LinkMetricsQueryOptionsTlv : public Tlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void)
+    {
+        SetType(kLinkMetricsQueryOptions);
+        SetLength(0);
+    }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval TRUE   If the TLV appears to be well-formed.
+     * @retval FALSE  If the TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const
+    {
+        return GetLength() >= sizeof(LinkMetricsTypeIdFlags) &&
+               GetLength() <= sizeof(LinkMetricsTypeIdFlags) * kMaxTypeIdFlagsCount;
```

Comment:
Same here. TLV should be considered valid is its length is larger than something.